cachito/workers/prune_archives.py

```diff
@@ -0,0 +1,233 @@
+    ref: str
+
+    @classmethod
+    def from_path(cls, path: Path) -> "_ParsedArchive":
+        repo_name = path.parent.relative_to(ARCHIVE_DIR).as_posix()
+        ref = path.name[:40]
+        return cls(path, repo_name, ref)
+
+
+class _ResolvedArchive(NamedTuple):
+    """A source archive matched to the most recent request for it."""
+
+    path: Path
+    created: datetime
+    latest_request_id: int
+
+
+@app.callback()
+def configure_logging(verbose: bool = False):
+    """Configure logging for the app."""
+    log_level = logging.DEBUG if verbose else logging.INFO
+    handler = logging.StreamHandler()
+    handler.setFormatter(logging.Formatter(LOG_FORMAT))
+    log.setLevel(log_level)
+    log.addHandler(handler)
+
+
+def _get_latest_request(archive: _ParsedArchive) -> Optional[dict[str, Any]]:
+    """
+    Find the latest request matching the _ParsedArchive via the API.
+
+    Return None if no matching request is found.
+    """
+    url = f"{config.cachito_api_url.rstrip('/')}/requests/latest"
+    params = {
+        "repo_name": archive.repo_name,
+        "ref": archive.ref,
+    }
+
+    try:
+        response = session.get(url, params=params, timeout=config.cachito_api_timeout)
+        response.raise_for_status()
+    except requests.HTTPError:
+        if response.status_code == 404:
+            return None
+        log.error(
+            "The request to %s failed with the status code %d and the following text: %s",
+            url,
+            response.status_code,
+            response.text,
+        )
+        raise NetworkError("Failed to query the cachito API")
+    except requests.RequestException:
+        msg = f"The connection failed when querying {url}"
+        log.exception(msg)
+        raise NetworkError(msg)
+
+    return response.json()
+
+
+def _get_parsed_source_archives(archive_dir: Path) -> Generator[_ParsedArchive, None, None]:
+    """Return a _ParsedArchive for each source archive in ARCHIVE_DIR."""
+
+    def is_valid_archive_filename(filename: str) -> bool:
+        """Archive filename should match <sha1 hash>-<(optional)with-submodules>.tar.gz."""
+        return re.match(ARCHIVE_PATTERN, filename) is not None
+
+    for path in archive_dir.rglob("*.tar.gz"):
+        if path.is_file() and is_valid_archive_filename(path.name):
+            yield _ParsedArchive.from_path(path)
+        else:
+            log.debug("%s does not appear to be a source archive.", path)
+
+
+def _resolve_source_archive(parsed_archive: _ParsedArchive) -> Optional[_ResolvedArchive]:
+    """Return a _ResolvedArchive if a matching request is found via the API."""
+    latest_request = _get_latest_request(parsed_archive)
+    if latest_request is None:
+        log.debug("Archive %s could not be resolved via the API.", parsed_archive.path)
+        return None
+
+    return _ResolvedArchive(
+        parsed_archive.path,
+        datetime.strptime(latest_request["created"], "%Y-%m-%dT%H:%M:%S.%f").replace(
```

Comment:
Are the request timestamps already UTC adjusted or do we need to use the same treatment here as with user CLI input? I'm just curious since I don't see a 'Z' in the unit test timestamps and I'm too lazy to go and check the deployment for real data :).
It's still a nitpick though, so even if it weren't no big deal since I doubt the time shift would make a significant difference in terms of archive cleanup anyway.

Reply:
Yeah, the request timestamps returned from the DB all should be UTC already. They're the values formatted by datetime.isoformat() in the first commit here

Reply:
@taylormadore can you specifically request review from me so that my approval counts?